librbd: deep copy optionally support flattening cloned image

trociny opened the pull request:
Fix: http://tracker.ceph.com/issues/22787
Signed-off-by: Mykola Golub <mgolub@suse.com>

trociny commented:
@dillaman Initially the idea was just to implement "flattening" support, but when I added tests I realized that the current deep copy for a cloned image, without "flattening", works incorrectly for the case like below (see TestDeepCopy.Clone_LargerDstObjSize):

1) fill an image with data
2) create a clone
3) write to the first object
4) deep copy the image to a destination with larger object size.

On the step 3 the copyup is triggered for the first child object. The second child object is empty. When we deep copy, the child objects 1 and 2 are copied to the destination object 1. When we read from this object, it returns zero for the region corresponding to the source object 2 (it does not try to read from the parent because the object exists).

If my interpretation is correct I think we need to detect such cases in `deep_copy::CopyObjectRequest` and do "flattening". That is why I did not create a separate fix for this issue, but add the code here, to the flattening feature.

My current implementation is not cleaned and not complete though -- the function that should detect if "flattening" is needed (`is_copy_from_parent_required`) currently just returns true.

Do you think my approach is correct in general and how do you suggest then to detect properly if copy from parent is required?

dillaman commented:
@trociny Yeah, it makes sense to me. I think if you get an ```-ENOENT``` on a source object when computing the diffs (i.e. the object doesn't exist), you should read-from-parent on the destination object's extents within the parent and associate the write with the empty ```snapc``` in the destination.

trociny commented:
@dillaman I have pushed an implementation of "recursive copy from parent", though I still observing some failures I need to check. But here are some issues I found and would like to discuss:

1) When deep copy detected that the object size truncated to zero we replaced truncate op with object remove, which was wrong in a cloned image case, because we would need to leave empty object to hide the parent. I have added a patch that fixes this, though it probably needs improvement: we may still use remove op when there is no parent.

2) When `api::Image::deep_copy` is run for a clone, it creates the destination clone, but does not check that the source clone size may be different than parent size (the image was shrinked after clone). As a result the deep copy fails in copy snapshots -> set header: it detects the parent overlap does not match and tries to set parent, which fails with EEXIST. I added the patch that resize the destination image after the clone request, if it is needed. On the other hand we could probably detect this in SetHeaderRequest and make it work properly for such a case.

3) I have added a simple test  TestDeepCopy.CloneDiscard that illustrace another failure case:
- write to an image
- clone
- discard the object with write
On discard we do "create+truncate" for this object, so it hides the parent. But when deep copying the object `calc_snap_set_diff` returns that it does not exist, so we don't create an empty object on destination. I am not sure how to fix this. It look like we would need to run read for all such objects to check if they really exist.

dillaman commented:
@trociny 

> I have added a simple test TestDeepCopy.CloneDiscard that illustrace another failure case:
> 
> write to an image
> clone
> discard the object with write
> On discard we do "create+truncate" for this object, so it hides the parent. But when deep copying the object calc_snap_set_diff returns that it does not exist, so we don't create an empty object on destination. I am not sure how to fix this. It look like we would need to run read for all such objects to check if they really exist.

In this case, if all your source objects are empty, you would just need to issue a create+truncate if there is a parent overlap, correct?

dillaman commented on src/librbd/deep_copy/ObjectCopyRequest.cc:
```diff
@@ -144,7 +180,8 @@ void ObjectCopyRequest<I>::send_read_object() {
     }
 
-    send_write_object();
+    compute_parent_object_extents();
```
You should only need to read from the parent on the first snapshot (or HEAD revision) that you are copying. Should the ```read_from_parent``` logic be moved to ```compute_read_ops``` for clarity? If it's ```m_snap_set``` does not contain any extents for ```start_src_snap_id == 0``` and the object overlaps w/ the parent for ```end_src_snap_id```, you will have to read from the parent and you can execute that before running ```send_read_object```.

When you do read from the parent, if the data is all zero, you will have to add a write-op for ```create```, otherwise it's just a write-op for the data read from the parent. In the future, we can improve our internal IO read path to keep IO sparse, but for now I'd say don't worry about trying to keep the destination sparse if some parent data exists.

dillaman commented on src/tools/rbd/ArgumentTypes.cc:
```diff
@@ -367,6 +367,11 @@ void add_export_format_option(boost::program_options::options_description *opt)
 
+void add_flatten_option(boost::program_options::options_description *opt) {
+  opt->add_options()
+    (IMAGE_FLATTEN.c_str(), po::bool_switch(), "copy clone parent data");
```
"Fill clone with parent data (make it independent)"

dillaman commented on src/include/rbd/librbd.h:
```diff
@@ -206,7 +206,8 @@ enum {
   RBD_IMAGE_OPTION_FEATURES_CLEAR = 9,
-  RBD_IMAGE_OPTION_DATA_POOL = 10
+  RBD_IMAGE_OPTION_DATA_POOL = 10,
+  RBD_IMAGE_OPTION_FLATTEN = 11,
```
Since this only applies to deep-copy, perhaps create/copy/clone should throw an error if flatten is specified.

dillaman commented on src/librbd/api/Image.cc:
```diff
@@ -302,6 +310,11 @@ int Image<I>::deep_copy(I *src, librados::IoCtx& dest_md_ctx,
 
+  if (dst_size != src_size) {
+    NoOpProgressContext no_op;
+    r = dest->operations->resize(src_size, true, no_op);
```
Can we just adjust the check in ```SetHeadRequest::send_remove_parent``` to remove the parent if the overlap doesn't match?

trociny replied:
Probably call `cls::set_size` instead

trociny replied:
Oh, we already calls set_size as the first step. So it should update the overlap. It looks like we just don't need to call remove/set parent when overlap does not match. I will try this.

dillaman commented on src/librbd/deep_copy/ObjectCopyRequest.cc:
```diff
@@ -965,15 +965,9 @@ void ObjectCopyRequest<I>::compute_zero_ops() {
-              .emplace_back(COPY_OP_TYPE_TRUNC, 0, z.get_start(), 0);
-            ldout(m_cct, 20) << "COPY_OP_TYPE_TRUNC " << z.get_start() << dendl;
-          }
+          m_write_ops[src_snap_seq]
```
Nit: perhaps just put a ```TODO``` since if you are flattening, you can send the remove.

trociny replied:
@dillaman 
> In this case, if all your source objects are empty, you would just need to issue a create+truncate if there is a parent overlap, correct?

That means that if we have a large parent image with only one object used, clone it, then deep copy will create many empty objects (for every nonexistent object in parent)?

dillaman replied:
> On discard we do "create+truncate" for this object, so it hides the parent. But when deep copying the object calc_snap_set_diff returns that it does not exist

Actually -- thinking about this, if the original clone had to issue a ```create+truncate```, why would ```calc_snap_set_diff``` return ```-ENOENT```? The object should exist, but it should be size zero. When deep-copying this object, if this is a clone w/ parent overlap, you would just need to preserve the ```create```, right?

dillaman replied:
Hmm -- sounds like a bug in ```calc_snap_set_diff```. If retrieving the ```snap_set_t``` worked, it means the object exists. If the object doesn't have any clones, it means the object exists throughout the full history of the object.

trociny commented on src/librbd/deep_copy/ObjectCopyRequest.cc:
```diff
@@ -58,9 +58,45 @@ ObjectCopyRequest<I>::ObjectCopyRequest(I *src_image_ctx, I *dst_image_ctx,
+  RWLock::RLocker snap_locker(m_parent->snap_lock);
+  RWLock::RLocker parent_locker(m_parent->parent_lock);
+  m_parent_child_snap_id = m_parent->snap_id;
+  m_parent = m_parent->parent;
```
Do you mean I should use `librbd::io::ImageRequestWQ::aio_read` instead of librados object read, when reading from parent?

trociny commented:
@dillaman updated

trociny commented on src/librbd/deep_copy/ObjectCopyRequest.cc:
```diff
@@ -105,7 +105,10 @@ void ObjectCopyRequest<I>::handle_list_snaps(int r) {
         e.noent = true;
       }
     }
+    m_exists = false;
```
It looks like I can just skip calling `compute_read_ops` in this case.

trociny commented:
@dillaman thanks! Updated.

dillaman commented:
@trociny Can you rebase this against master?

trociny commented:
> Can you rebase this against master?

Sure. I am not sure though how to do something like this on github though:

> liewegas changed the base branch from master  to mimic  13 days ago

dillaman commented:
I suspect that's why my usual merge failed since it started creating and deleting random files all over the Ceph tree. I just cherry-picked your individual commits as a temporary workaround.

trociny commented:
@dillaman I pushed my branch rebased against master. I don't know how to change the base branch from mimic to master on github.

@liewegas do you have an idea?

dillaman commented:
@trociny Didn't see that this PR was reassigned to the mimic branch -- all fixed.
